src/merkle_tree.rs

```diff
@@ -0,0 +1,206 @@
+use hmac_sha256::Hash;
+
+pub struct MerkleTree {
+    pub root: Option<[u8; 32]>,
+    pub leaves: Vec<[u8; 32]>,
+}
+
+impl MerkleTree {
+    /// Create a new empty MerkleTree instance.
+    pub fn new() -> Self {
+        Self {
+            root: None,
+            leaves: Vec::new(),
+        }
+    }
+
+    /// Create a new MerkleTree from the provided items.
+    /// Each item should be representable as bytes.
+    /// It returns a `MerkleTree` instance with the leaf hashes and the Merkle root.
+    pub fn build<T: AsRef<[u8]>>(items: &[T]) -> Self {
+        if items.is_empty() {
+            return Self::new();
+        }
+
+        let leaves: Vec<[u8; 32]> = items.iter().map(|item| Self::hash(item.as_ref())).collect();
+
+        Self {
+            root: Some(Self::merkle_root(leaves.clone())),
+            leaves,
+        }
+    }
+
+    /// Computes the parent hash for the concatenation of the provided left and right hashes.
+    fn merkle_parent(left: &[u8; 32], right: &[u8; 32]) -> [u8; 32] {
+        let concat = [left.as_slice(), right.as_slice()].concat();
+
+        Self::hash(&concat)
+    }
```

Comment:
`left` and `right` are always consecutive in memory, so we may save ourselves an allocation by using [as_flattened](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_flattened):
```suggestion
    fn merkle_parent(children: &[[u8; 32]; 2]) -> [u8; 32] {
        Self::hash(children.as_flattened())
    }
```

Reply:
Based on your suggestion, I propose this:

```
    /// Computes the parent hash for the concatenation of the children hashes.
    fn merkle_parent(children: &[[u8; 32]]) -> [u8; 32] {
        Self::hash(children.as_flattened())
    }
```

for two reasons:
1. The method is easier to use, I don't have to check the length of the `children` slice equals 2 each time I want to use it.
2. It actually makes the method more generic, as it's not tied to having two children. It can be used to compute the merkle parent of an arbitrary amount of children hashes.

What do you think?